My Jetpack: Connect standalone plugin menu actions

lhkowalski opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->
<!-- a12s: If you have an expected version that you're aiming for the PR to add, please use the Milestone field to communicate it. If you leave it blank, that indicates there isn't a preference. -->

Relates to #29901.

## Proposed changes:
<!--- Explain what functional changes your PR includes -->
* Change the `use-product` hook to export the standalone plugin installation action
* Create function to handle the call for the installation and pass it to the `ProductCard`
* Fix menu visibility control to consider the standalone plugin status
* Enable the action menu for VideoPress
* Track events related to install and activate standalone plugin

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->

## Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->

## Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

### Testing the installation + activation of a standalone product

* Spin a new JN site with only Jetpack
* Connect the site
* Go to `Jetpack > My Jetpack` and check the VideoPress card; it should show as inactive since we don't activate it yet; it will also show the three dots to open the actions menu:

<img width="369" alt="Screen Shot 2023-04-18 at 18 44 38" src="https://user-images.githubusercontent.com/6760046/232912126-010fa9a7-f69d-4921-ad4d-e508ec37de97.png">

* Click on the action menu icon, on the top right corner of the VideoPress card
* You should see the "Install Plugin" action available:

<img width="363" alt="Screen Shot 2023-04-18 at 18 45 44" src="https://user-images.githubusercontent.com/6760046/232912255-acbfb320-921e-4902-ad78-511d93734162.png">

* Click on "Install Plugin"
* The card will switch to a loading state and then the plugin will be installed and activated
* After the loading state finishes, refresh the page to get the card with the new status; there is a known issue that the install standalone endpoint is returning a stale product status after the installation; I will fix it on a follow up;
* After the refresh, the card is now shown as active and the actions menu only shows the "Manage" option:

<img width="368" alt="Screen Shot 2023-04-18 at 18 48 29" src="https://user-images.githubusercontent.com/6760046/232912735-307e0d4c-6cbb-4821-85ce-a03009cce9a4.png">

* Go to `Plugins > Installed Plugins` and confirm that the VideoPress plugin is now installed:

<img width="349" alt="Screen Shot 2023-04-17 at 18 38 56" src="https://user-images.githubusercontent.com/6760046/232616754-f6ad5412-dd72-46ef-968d-6e41e2cf1027.png">

### Testing the activation of a standalone product

* We are going to disable the plugin to check the action menu again
* On the same test site, go to `Plugins > Installed Plugins` and disable the VideoPress plugin
* Go to `Jetpack > My Jetpack` and confirm that the VideoPress card still show as active (since the module on Jetpack got enable on the prior step):

<img width="358" alt="Screen Shot 2023-04-18 at 18 50 02" src="https://user-images.githubusercontent.com/6760046/232913024-aca77500-e085-4f1b-abf9-78d50b4a60b9.png">

* Click on the action menu icon, on the top right corner of the VideoPress card
* You should see the "Manage" (since the product is active) and "Activate Plugin" (since the plugin is installed but not active) actions available:

<img width="361" alt="Screen Shot 2023-04-18 at 18 51 04" src="https://user-images.githubusercontent.com/6760046/232913191-d1a45bbd-19a6-4b30-a48c-4521ce786e7a.png">

* Click on "Activate Plugin"
* The card will switch to a loading state and then the plugin will be activated on the backend
* After the loading state finishes, the card still shows as active and the actions menu only shows the "Manage" option:

<img width="367" alt="Screen Shot 2023-04-18 at 18 51 37" src="https://user-images.githubusercontent.com/6760046/232913331-10bc508b-0b6c-4162-8056-59a44f45c471.png">

* Go to `Plugins > Installed Plugins` and confirm that the VideoPress plugin is now active again:

<img width="349" alt="Screen Shot 2023-04-17 at 18 38 56" src="https://user-images.githubusercontent.com/6760046/232616754-f6ad5412-dd72-46ef-968d-6e41e2cf1027.png">

lhkowalski committed:
Fix casing of vars on product selector

lhkowalski committed:
Get standalone plugin info and pass it to ProductCard

lhkowalski committed:
Add action to dispatch plugin installation and activation on use-product hook

lhkowalski committed:
Connect action to install and activate plugin

lhkowalski committed:
Add support for the showMenu prop to control the menu presence on a product basis

lhkowalski committed:
Enable action menu presence for VideoPress product

lhkowalski committed:
Prevent showing the menu when there is no options to show or the status is absent

lhkowalski committed:
Add changelog file

lhkowalski committed:
Include Tracks events for the standalone actions

lhkowalski committed:
Update standalone action icons

lhkowalski committed:
Only show the Manage menu option when the product is active

lhkowalski committed:
Simplify menu generation for manage menu

lhkowalski committed:
Simplify menu generation for standalone actions

renatoagds commented on projects/packages/my-jetpack/_inc/components/product-card/index.jsx:
```diff
@@ -34,6 +35,30 @@ const Menu = ( {
+			onClick: onActivate,
+			icon: check,
+		} );
+	}
```
I think it is better not to manipulate the prop directly. We're changing the original array, and if we need to rely on it at some moment, it will be another one. And this way, we change the layout, not including the divider between "actions" and "install/activate." If you prefer to keep using array, better to split between custom items and default items, but I vote to keep the default items declarative.

lhkowalski replied:
Good point!

About the divider, I ended up having the menu showing it above the install option but with nothing above the divider, I think when the module is not active and the plugin is not installed.

I will move back to the declarative format and work on a small condition to show/not show the divider.

lhkowalski replied:
Changed the code back to the declarative format and also added a check to show the divider.

renatoagds commented on projects/packages/my-jetpack/_inc/components/connected-product-card/index.jsx:
```diff
@@ -40,6 +41,30 @@ const ConnectedProductCard = ( { admin, slug, children } ) => {
 
+	/*
+	 * Redirect to connection page if the user is not connected
+	 */
```
I vote to remove the installation option if the user isn't connected. Strange experience to be moved for connection after clicking on install, and in the end, nothing will be installed. We have a big notice at the top warning the user that isn't connected.

lhkowalski replied:
Technically it's not impossible to trigger the installation for a disconnected site, but I agree that it may be a sub-optimal experience. I will try removing the option to see how things show up.

lhkowalski replied:
Removed the connection redirect from the code and added a check for the `error` status, but I think it's not enough yet. I will figure out a way to have the connection state available on the menu to include it on the button condition.

lhkowalski replied:
Done. I injected the connection state into the product card and use it to decide between showing or not the menu. For me it does make sense to hide the menu everytime there is a connection problem, since it will not show anything that could work disconnected and will not help fix the connection issue, at least for now. Do you agree?

lhkowalski committed:
Move default menus back to a declaritive format for better readability

lhkowalski committed:
Bring back standalone options divider

lhkowalski committed:
Only show the standalone menu options when there is no connection error

lhkowalski committed:
Remove check for connection status on the standalone install handler

Since we removed the install button from menu when there is a connection error, we don't need the check for the connection status while installing.

lhkowalski committed:
Fix empty menu bug when the product status is error

lhkowalski committed:
Only show the actions menu when the connection is set properly
